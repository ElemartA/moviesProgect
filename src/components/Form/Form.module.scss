.wrap {
  width: 470px;
  min-height: 600px;
  height: auto;
  margin: 0 auto;
  background:  linear-gradient(
    rgba(188, 188, 212, 0.8),
    rgba(110, 105, 132, 0.688));
  font-family: "Cormorant Garamond", serif;
  padding-bottom: 5px;

  -webkit-box-shadow: 0px 0px 24px 8px rgba(120, 37, 219, 0.12);
  -moz-box-shadow: 0px 0px 24px 8px rgba(120, 37, 219, 0.12);
  box-shadow: 0px 0px 24px 8px rgba(120, 37, 219, 0.12);

  .form {
    color: #fff;
    margin-left: 1rem;
    font-size: 20px;
    display: block;
    margin-left: 0;
    margin-bottom: 5px;
    :nth-child(1n) {
      margin-bottom: 15px;
      display: block;
      width: 93%;
      margin: 0;
      text-align: left;
      padding: 10px;
      border: 0;
      border-radius: 20px;
      opacity: 0.85;
      font-size: 18px;
      font-weight: 600;
    }
    .label {
      margin-bottom: 15px;
      
      .name {
        margin-top: 1rem;
        background-color: lightgray;
        height: 25px;
        font-size: 20px;
        padding-left: 10px;
        font-family: "Cormorant Garamond", serif;
        font-weight: 600;
      }
    }

    .switch {
      position: relative;
      display: inline-block;
      width: 8px;
      height: 0px;
      margin-top: 13px;
      margin-right: 21px;
    }

    /* Hide default HTML checkbox */
    .switch input {
      display: none;
    }

    /* The slider */
    .slider {
      position: absolute;
      cursor: pointer;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      background-color: #ccc;
      -webkit-transition: 0.4s;
      transition: 0.4s;
     
    }

    .slider:before {
      position: absolute;
      content: "";
      height: 26px;
      width: 26px;
      left: 4px;
      bottom: 4px;
      background-color: white;
      -webkit-transition: 0.4s;
      transition: 0.4s;
    }

    input:checked + .slider {
      background-color: #7b2dce;
    }

    input:focus + .slider {
      box-shadow: 0 0 1px #7b2dce;
    }

    input:checked + .slider:before {
      -webkit-transform: translateX(26px);
      -ms-transform: translateX(26px);
      transform: translateX(26px);
    }

    /* Rounded sliders */
    .slider.round {
      border-radius: 34px;
      margin-top: -14px;
      width: 60px;
     
    }

    .slider.round:before {
      border-radius: 50%;
    }

    .checkbox {
      display: flex;
      margin: 0;
      flex-direction: row;
      justify-content: flex-start;
      align-items: flex-start;
    }
    .check {  
      display: flex;   
      .check__input {
       flex: 1 2 0px;
       margin-top: 5px;
    
      }
      .check__label {
        flex: 2 1 300px;
        line-height: 0px;
      }
    }

    .submit {
      font-family: "Cormorant Garamond", serif;
      font-weight: 600;
      width: 100px;
      height: 30px;
      font-size: 15px;
      margin-top: 10px;
      margin-left: 350px;
      line-height: 5px;
      font-size: 17px;
      font-weight: 700;
      color: black;
      opacity: 1;
      cursor: pointer;
    }
  }
}
