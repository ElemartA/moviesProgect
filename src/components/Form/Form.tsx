import React from "react";
import MiniCard from "../MiniCard";
import s from "./Form.module.scss";
import { useForm } from "react-hook-form";
import { FormInputs } from "../../types/formInputsType";
import { useAppDispatch, useAppSelector } from "../../hooks/redux";
import { formSlice } from "../../store/reducers/FormSlice";

const Form = () => {
  const { miniCardValue } = useAppSelector((state) => state.formReducer);
  const { crete_miniCard } = formSlice.actions;
  const dispatch = useAppDispatch();
  const {
    register,
    formState: { errors, isValid },
    handleSubmit,
    reset,
  } = useForm<FormInputs>({
    mode: "all",
  });

  const submit = (data: FormInputs) => {
    let value = "";
    let agree = "";
    data.checkbox === true
      ? (value = "хочу получать уведомления")
      : (value = "не хочу получать уведомления");

    data.switch === true
      ? (agree = "согласен на обработку данных")
      : (agree = "не согласен на обработку данных");

    if (data.file.length !== 0) {
      dispatch(
        crete_miniCard([
          ...miniCardValue,
          {
            name: data.firstName,
            date: data.lastName,
            country: data.country,
            check: value,
            check2: agree,
            imgSrc: URL.createObjectURL(data.file[0] as Blob),
          },
        ])
      );
    }

    reset();
  };

  return (
    <div className={s.wrap}>
      <form onSubmit={handleSubmit(submit)} className={s.form}>
        <label>
          Имя:
          <input
            placeholder="Имя"
            data-testid="name"
            {...register("firstName", {
              required: "Поле обязательно к заполнению",
              minLength: {
                value: 5,
                message: "Длинна не менее 5 символов",
              },
            })}
          ></input>
        </label>
        <div style={{ height: 40, color: "red" }}>
          {errors?.firstName && <p>{errors?.firstName?.message || "Error!"}</p>}
        </div>
        <label>
          Фамилия:
          <input
            placeholder="Фамилия"
            {...register("lastName", {
              required: "Поле обязательно к заполнению",
              minLength: {
                value: 5,
                message: "Длинна не менее 5 символов",
              },
            })}
          ></input>
        </label>
        <div style={{ height: 40, color: "red" }}>
          {errors?.lastName && <p>{errors?.lastName?.message || "Error!"}</p>}
        </div>
        <label>
          <input
            data-testid="date"
            type="date"
            {...register("date", {
              required: "Поле обязательно к заполнению",
            })}
          />
        </label>
        <div style={{ height: 40, color: "red" }}>
          {errors?.date && <p>{errors?.date?.message || "Error!"}</p>}
        </div>
        <label htmlFor="country">
          Страна:
          <select
            {...register("country", {
              required: "Поле обязательно к заполнению",
            })}
          >
            <option>Belarus</option>
            <option>Russia</option>
            <option>Ukraine</option>
          </select>
          <div style={{ height: 40, color: "red" }}>
            {errors?.country && <p>{errors?.country?.message || "Error!"}</p>}
          </div>
        </label>
        <label>
          Загрузите фото:
          <input
            {...register("file", {
              required: "Загрузите фото",
            })}
            type="file"
          />
          <div style={{ height: 40, color: "red" }}>
            {errors?.file && <p>{errors?.file?.message || "Error!"}</p>}
          </div>
        </label>
        <div className={s.checkbox}>
          <label className={s.switch}>
            <input {...register("checkbox")} type="checkbox"></input>
            <span className={`${s.slider} ${s.round}`}></span>
          </label>
          <label style={{ paddingLeft: 23 }}>
            не хочу/хочу получать уведомления об акциях
          </label>
        </div>
        <div className={s.check}>
          <input
            className={s.check__input}
            type="checkbox"
            {...register("switch")}
          />
          <label className={s.check__label}>
            Я согласен на обработку данных
          </label>
        </div>
        <input
          disabled={!isValid}
          className={s.submit}
          type="submit"
          value="Отправить"
        />
      </form>
      <MiniCard value2={miniCardValue} />
    </div>
  );
};

export default Form;
