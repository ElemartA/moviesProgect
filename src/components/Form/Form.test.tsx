import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Form from "./Form";

describe("Form", () => {
  it("renders Form component", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Form />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
    screen.debug();
    expect(screen.getByRole("combobox")).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeDisabled();
    expect(screen.queryByText(/Имя/i)).toBeInTheDocument();
    expect(screen.queryByText(/Страна:/i)).toBeInTheDocument();
    expect(screen.queryByText(/Главная/i)).toBeNull();

    const btn = screen.getByText(/Отправить/i);
    expect(btn).toBeDisabled();
    const input = screen.getByTestId("name");
    fireEvent.input(input);
    fireEvent.click(btn);
  });
  it("INPUT EVENT", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Form />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
    screen.debug();
    const input = screen.getByTestId("name");
    const inputDate = screen.getByTestId("date");
    expect(input).toContainHTML("");
    expect(inputDate).toContainHTML("");
    fireEvent.input(input, {
      target: { value: "тестовая строка" },
    });
    expect(input).toHaveValue("тестовая строка");
  });
});
