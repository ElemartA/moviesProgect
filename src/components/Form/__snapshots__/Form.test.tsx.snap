// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form INPUT EVENT 1`] = `
<DocumentFragment>
  <div
    class="wrap"
  >
    <form
      class="form"
    >
      <label>
        Имя:
        <input
          data-testid="name"
          name="firstName"
          placeholder="Имя"
        />
      </label>
      <div
        style="height: 40px; color: red;"
      />
      <label>
        Фамилия:
        <input
          name="lastName"
          placeholder="Фамилия"
        />
      </label>
      <div
        style="height: 40px; color: red;"
      />
      <label>
        <input
          data-testid="date"
          name="date"
          type="date"
        />
      </label>
      <div
        style="height: 40px; color: red;"
      />
      <label
        for="country"
      >
        Страна:
        <select
          name="country"
        >
          <option>
            Belarus
          </option>
          <option>
            Russia
          </option>
          <option>
            Ukraine
          </option>
        </select>
        <div
          style="height: 40px; color: red;"
        />
      </label>
      <label>
        Загрузите фото:
        <input
          name="file"
          type="file"
        />
        <div
          style="height: 40px; color: red;"
        />
      </label>
      <div
        class="checkbox"
      >
        <label
          class="switch"
        >
          <input
            name="checkbox"
            type="checkbox"
          />
          <span
            class="slider round"
          />
        </label>
        <label
          style="padding-left: 23px;"
        >
          не хочу/хочу получать уведомления об акциях
        </label>
      </div>
      <div
        class="check"
      >
        <input
          class="check__input"
          name="switch"
          type="checkbox"
        />
        <label
          class="check__label"
        >
          Я согласен на обработку данных
        </label>
      </div>
      <input
        class="submit"
        disabled=""
        type="submit"
        value="Отправить"
      />
    </form>
    <div
      class="wrap"
      data-testid="miniCard"
    >
      <div />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Form renders Form component 1`] = `
<DocumentFragment>
  <div
    class="wrap"
  >
    <form
      class="form"
    >
      <label>
        Имя:
        <input
          data-testid="name"
          name="firstName"
          placeholder="Имя"
        />
      </label>
      <div
        style="height: 40px; color: red;"
      />
      <label>
        Фамилия:
        <input
          name="lastName"
          placeholder="Фамилия"
        />
      </label>
      <div
        style="height: 40px; color: red;"
      />
      <label>
        <input
          data-testid="date"
          name="date"
          type="date"
        />
      </label>
      <div
        style="height: 40px; color: red;"
      />
      <label
        for="country"
      >
        Страна:
        <select
          name="country"
        >
          <option>
            Belarus
          </option>
          <option>
            Russia
          </option>
          <option>
            Ukraine
          </option>
        </select>
        <div
          style="height: 40px; color: red;"
        />
      </label>
      <label>
        Загрузите фото:
        <input
          name="file"
          type="file"
        />
        <div
          style="height: 40px; color: red;"
        />
      </label>
      <div
        class="checkbox"
      >
        <label
          class="switch"
        >
          <input
            name="checkbox"
            type="checkbox"
          />
          <span
            class="slider round"
          />
        </label>
        <label
          style="padding-left: 23px;"
        >
          не хочу/хочу получать уведомления об акциях
        </label>
      </div>
      <div
        class="check"
      >
        <input
          class="check__input"
          name="switch"
          type="checkbox"
        />
        <label
          class="check__label"
        >
          Я согласен на обработку данных
        </label>
      </div>
      <input
        class="submit"
        disabled=""
        type="submit"
        value="Отправить"
      />
    </form>
    <div
      class="wrap"
      data-testid="miniCard"
    >
      <div />
    </div>
  </div>
</DocumentFragment>
`;
